plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/command/CommandManager.java

```diff
@@ -128,13 +139,55 @@ public void apply(MachineProcessDto process) throws OperationException {
      *
      * @see CommandPropertyValueProvider
      */
-    public String substituteProperties(final String commandLine) {
-        String cmdLine = commandLine;
+    public Promise<String> substituteProperties(String commandLine) {
+        Promise<String> promise = Promises.resolve(null);
+        CommandLineContainer commandLineContainer = new CommandLineContainer(commandLine);
+        return replaceParameters(promise, commandLineContainer, commandPropertyValueProviderRegistry.getProviders().iterator());
+    }
 
-        for (CommandPropertyValueProvider provider : commandPropertyValueProviderRegistry.getProviders()) {
-            cmdLine = cmdLine.replace(provider.getKey(), provider.getValue());
+    private Promise<String> replaceParameters(Promise<String> promise,
+                                              CommandLineContainer commandLineContainer,
+                                              Iterator<CommandPropertyValueProvider> iterator) {
+        if (!iterator.hasNext()) {
+            return promise;
         }
 
-        return cmdLine;
+        final CommandPropertyValueProvider provider = iterator.next();
+
+        Promise<String> derivedPromise = promise.thenPromise(proceedRefactoringMove(commandLineContainer, provider));
+
+        return replaceParameters(derivedPromise, commandLineContainer, iterator);
+    }
+
+    private Function<String, Promise<String>> proceedRefactoringMove(final CommandLineContainer commandLineContainer,
+                                                                     final CommandPropertyValueProvider provider) {
+        return new Function<String, Promise<String>>() {
+            @Override
+            public Promise<String> apply(String arg) throws FunctionException {
+                return provider.getValue().thenPromise(new Function<String, Promise<String>>() {
+                    @Override
+                    public Promise<String> apply(String arg) throws FunctionException {
+                        commandLineContainer.setCommandLine(commandLineContainer.getCommandLine().replace(provider.getKey(), arg));
+                        return Promises.resolve(commandLineContainer.getCommandLine());
+                    }
+                });
+            }
+        };
+    }
+
+    private class CommandLineContainer {
```

Comment:
I can't use String object, because this field should be private and I also need to replace this field  in the process of building the command.